### PR TITLE
[FEATURE]: Optimistic Updates

### LINKED ISSUE #330: [FEATURE]: Optimistic Updates
## Summary
In live version exists some delay performing actions like adding votes, adding or edit comments, deleting cards, etc.

### Motivation

Add optimistic updates to improve the user experience. 

### Add optimistic updates:

- [x] Add/Delete card
- [x] Merge cards
- [ ] Add/Remove vote
- [x] Add/Edit/Delete comment

Feel free to add more tasks if necessary

### PR DESCRIPTION
**Pull request by dvpfran**
Relates to #330

People involved:
@joaosantos99 @geomarb 

This pull request closes #330

### CODE REVIEW

File: frontend/src/hooks/useComments.tsx [hunk 1]
```diff
@@ -1,43 +1,151 @@
...
+							return {
+								board: {
+									...old.board,
+									columns: newBoard.columns
+								}
+							};
+
+						return old;
+					}
+				);
```
Comment (dvpfran) [inline frontend/src/hooks/useComments.tsx:43]: Create a function with this code, it's being called multiple times. Something like `updateBoardColumns`

File: frontend/src/hooks/useComments.tsx [hunk 2]
```diff
@@ -1,43 +1,151 @@
...
+			}
+
+			return { previousBoard: board, data };
+		},
+		onSettled: (data) => {
 			queryClient.invalidateQueries(['board', { id: data?._id }]);
 		},
-		onError: () => {
+		onSuccess: () => {},
```
Comment (dvpfran) [inline frontend/src/hooks/useComments.tsx:51]: You can remove this line

File: frontend/src/hooks/useComments.tsx [hunk 3]
```diff
@@ -1,43 +1,151 @@
...
```
Comment (dvpfran) [inline frontend/src/hooks/useComments.tsx:97]: Same here about removing line

File: frontend/src/hooks/useComments.tsx [hunk 4]
```diff
@@ -1,43 +1,151 @@
...
+								board: {
+									...old.board,
+									columns: newBoard.columns
+								}
+							};
+
+						return old;
+					}
+				);
+			}
```
Comment (dvpfran) [inline frontend/src/hooks/useComments.tsx:136]: Same here about creating a function

File: frontend/src/hooks/useComments.tsx [hunk 5]
```diff
@@ -1,43 +1,151 @@
...
+
+			return { previousBoard: board, data };
+		},
+		onSettled: (data) => {
 			queryClient.invalidateQueries(['board', { id: data?._id }]);
 		},
-		onError: () => {
+		onSuccess: () => {},
```
Comment (dvpfran) [inline frontend/src/hooks/useComments.tsx:143]: Same here about removing line

File: frontend/src/hooks/useVotes.tsx [hunk 6]
```diff
@@ -1,57 +1,218 @@
...
+		onSettled: (boardDataFromApi) => {
+			queryClient.invalidateQueries(getBoardQueryKey(boardDataFromApi?._id));
+
+			if (boardDataFromApi && hasMaxVotesLimit(boardDataFromApi)) {
+				const remainingVotes = getRemainingVotes(boardDataFromApi, userId!);
+
+				const toastMessage = `You have ${remainingVotes} votes left`;
+
+				setToastState(buildToastMessage(toastMessage, ToastStateEnum.INFO));
 			}
```
Comment (dvpfran) [inline frontend/src/hooks/useVotes.tsx:186]: You can create a function for this and pass a parameter with the toast message string. You are using the same code at line 202. Only the toast message is different.

Comment (geomarb) [reply frontend/src/hooks/useVotes.tsx:186]: Great! I am doing it!

File: frontend/src/hooks/useVotes.tsx [hunk 7]
```diff
@@ -2,57 +2,257 @@ import { useMutation } from 'react-query';
...
+import BoardType from '../types/board/board';
 import { getRemainingVotes } from '../utils/getRemainingVotes';
 import useBoardUtils from './useBoardUtils';
 
+enum Action {
+	Add = 'add',
+	Remove = 'remove'
+}
+
+type QueryKeyType = (string | { id: string })[];
```
Comment (dsousa12) [inline frontend/src/hooks/useVotes.tsx:18]: Why you need this type?

**Review by dsousa12 — APPROVED** [review]

**Pull request merged**